render/vulkan/renderer.c

```diff
@@ -0,0 +1,1544 @@
+
+	// release render buffer after rendering
+	release_barriers[idx].sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+	release_barriers[idx].srcQueueFamilyIndex = renderer->dev->queue_family;
+	release_barriers[idx].dstQueueFamilyIndex = VK_QUEUE_FAMILY_FOREIGN_EXT;
+	release_barriers[idx].image = renderer->current_render_buffer->image;
+	release_barriers[idx].oldLayout = VK_IMAGE_LAYOUT_GENERAL;
+	release_barriers[idx].newLayout = VK_IMAGE_LAYOUT_GENERAL;
+	release_barriers[idx].srcAccessMask = VK_ACCESS_COLOR_ATTACHMENT_READ_BIT |
+		VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+	release_barriers[idx].dstAccessMask = 0u; // ignored anyways
+	release_barriers[idx].subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+	release_barriers[idx].subresourceRange.layerCount = 1;
+	release_barriers[idx].subresourceRange.levelCount = 1;
+	++idx;
+
+	vkCmdPipelineBarrier(pre_cb, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,
+		VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT | VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT,
+		0, 0, NULL, 0, NULL, barrier_count, acquire_barriers);
+
+	vkCmdPipelineBarrier(render_cb, VK_PIPELINE_STAGE_ALL_GRAPHICS_BIT,
+		VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT, 0, 0, NULL, 0, NULL,
+		barrier_count, release_barriers);
+
+	vkEndCommandBuffer(renderer->cb);
+
+	unsigned submit_count = 0u;
+	VkSubmitInfo submit_infos[2] = {0};
+
+	// No semaphores needed here.
+	// We don't need a semaphore from the stage/transfer submission
+	// to the render submissions since they are on the same queue
+	// and we have a renderpass dependency for that.
+	if (renderer->stage.recording) {
+		vkEndCommandBuffer(renderer->stage.cb);
+		renderer->stage.recording = false;
+
+		VkSubmitInfo *stage_sub = &submit_infos[submit_count];
+		stage_sub->sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+		stage_sub->commandBufferCount = 1u;
+		stage_sub->pCommandBuffers = &pre_cb;
+		++submit_count;
+	}
+
+	VkSubmitInfo *render_sub = &submit_infos[submit_count];
+	render_sub->sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+	render_sub->pCommandBuffers = &render_cb;
+	render_sub->commandBufferCount = 1u;
+	++submit_count;
+
+	VkResult res = vkQueueSubmit(renderer->dev->queue, submit_count,
+		submit_infos, renderer->fence);
+	if (res != VK_SUCCESS) {
+		wlr_vk_error("vkQueueSubmit", res);
+		return;
+	}
+
+	// sadly this is required due to the current api/rendering model of wlr
+	// ideally we could use gpu and cpu in parallel (_without_ the
+	// implicit synchronization overhead and mess of opengl drivers)
+	res = vkWaitForFences(renderer->dev->dev, 1, &renderer->fence, true,
+		UINT64_MAX);
+	if (res != VK_SUCCESS) {
+		wlr_vk_error("vkWaitForFences", res);
+		return;
+	}
+
+	++renderer->frame;
+	release_stage_allocations(renderer);
+
+	// destroy pending textures
+	wl_list_for_each_safe(texture, tmp_tex, &renderer->destroy_textures, destroy_link) {
+		wlr_texture_destroy(&texture->wlr_texture);
+	}
+
+	wl_list_init(&renderer->destroy_textures); // reset the list
```

Comment:
I think it's required since we don't ever remove textures from the destroy_textures list. We don't do it since this is the only place textures in the list are destroyed. I wouldn't be opposed to removing textures from the destroy_list in `vulkan_texture_destroy` instead but don't think its cleaner.